File: src/main/java/com/chattriggers/ctjs/modules/gui/ModulesGui.java
package com.chattriggers.ctjs.modules.gui;

import com.chattriggers.ctjs.minecraft.libs.ChatLib;
import com.chattriggers.ctjs.minecraft.libs.MathLib;
import com.chattriggers.ctjs.minecraft.libs.renderer.Renderer;
import com.chattriggers.ctjs.minecraft.libs.renderer.Text;
import com.chattriggers.ctjs.minecraft.objects.gui.GuiHandler;
import com.chattriggers.ctjs.modules.Module;
import com.chattriggers.ctjs.utils.config.Config;
import com.mojang.realmsclient.gui.ChatFormatting;
import net.minecraft.client.gui.GuiScreen;
import net.minecraft.client.renderer.GlStateManager;
import org.lwjgl.input.Mouse;

import java.io.IOException;
import java.util.ArrayList;

public class ModulesGui extends GuiScreen {
    private ArrayList<GuiModule> modules = new ArrayList<>();

    private int scrolled;
    private int maxScroll;

    public ModulesGui(ArrayList<Module> modules) {
        int i = 0;
        for (Module module : modules) {
            this.modules.add(new GuiModule(module, i));
            i++;
        }

        this.scrolled = 0;
        this.maxScroll = this.modules.size() * 110 + 10 - Renderer.screen.getHeight();

    }

    @Override
    public void drawScreen(int mouseX, int mouseY, float partialTicks) {
        super.drawScreen(mouseX, mouseY, partialTicks);

        GlStateManager.pushMatrix();

        this.maxScroll = this.modules.size() * 110 + 10 - Renderer.screen.getHeight();

        drawBackground(0);

        // draw scroll bar
        int scrollHeight = Renderer.screen.getHeight() - this.maxScroll;
        if (scrollHeight < 20) scrollHeight = 20;
        if (scrollHeight < Renderer.screen.getHeight()) {
            int scrollY = (int) MathLib.map(this.scrolled, 0, this.maxScroll, 10, Renderer.screen.getHeight() - scrollHeight - 10);
            Renderer.drawRect(
                    0xa0000000,
                    Renderer.screen.getWidth() - 5,
                    scrollY,
                    5,
                    scrollHeight
            );
        }

        for (GuiModule module : this.modules) {
            module.draw();
            module.checkHover(mouseX, mouseY);
        }

        GlStateManager.popMatrix();
    }

    @Override
    public void mouseClicked(int mouseX, int mouseY, int mouseButton) throws IOException {
        super.mouseClicked(mouseX, mouseY, mouseButton);

        if (mouseButton == 0) {
            for (GuiModule module : modules) {
                module.click();
            }
        }
    }

    @Override
    public void handleMouseInput() throws IOException {
        super.handleMouseInput();

        int i = Mouse.getEventDWheel();
        if (i == 0) return;

        if (i > 1) i = 1;
        if (i < -1) i = -1;

        if (isShiftKeyDown()) i *= 10;
        else i *= 20;

        this.scrolled -= i;
        if (this.scrolled > this.maxScroll)
            this.scrolled = this.maxScroll;
        if (this.scrolled < 0)
            this.scrolled = 0;
    }

    private class GuiModule {
        private Module module;
        private int i;

        private boolean isHovered;

        private int x;
        private int y;

        private String name;

        private GuiModule(Module module, int i) {
            this.module = module;
            this.i = i;

            this.isHovered = false;

            this.x = 20;
            this.y = getY(i);

            name = ChatLib.addColor(this.module.getMetadata().getDisplayName() == null
                    ? this.module.getName()
                    : this.module.getMetadata().getDisplayName());
        }

        private int getY(int i) {
            return i * 110 + 10 - scrolled;
        }

        private void click() {
            if (isHovered) {
                GuiHandler.getInstance().openGui(new ModuleGui(module));
            }
        }

        private void checkHover(int mouseX, int mouseY) {
            int width = Renderer.screen.getWidth() - 40;
            int height = 105;

            isHovered = (mouseX > x + width - Renderer.getStringWidth("show code >") - 2
                && mouseX < x + width - 2
                && mouseY > y + height - 12
                && mouseY < y + height - 2);
        }

        private void draw() {
            x = 20;
            y = getY(i);
            int width = Renderer.screen.getWidth() - 40;
            int height = 105;

            // background
            Renderer.drawRect(0x80000000, x, y, width, height);

            // name
            Renderer.text(name, x + 2, y + 2)
                    .setShadow(true).draw();

            // version
            if (this.module.getMetadata().getVersion() != null) {
                String version = ChatFormatting.GRAY  + "v" + this.module.getMetadata().getVersion();
                Renderer.text(version, x + width - Renderer.getStringWidth(version) - 2, y + 2)
                        .setShadow(true).draw();
            }

            // line break
            Renderer.drawRect(0xa0000000, x + 2, y+12, width - 4, 2);

            // description
            String description = (this.module.getMetadata().getDescription() == null)
                    ? "No description provided"
                    : this.module.getMetadata().getDescription();
            Text desc = Renderer.text(description, x + 2, y + 20).setWidth(width - 5).setMaxLines(6).draw();
            if (desc.exceedsMaxLines())
                Renderer.text("...", x + 2, y + 73).draw();


            // directory
            Renderer.text(ChatFormatting.DARK_GRAY + Config.getInstance().getModulesFolder().value + this.module.getName() + "/", x + 2, y + height - 12)
                    .setShadow(true).draw();

            // show code
            String finalShowCode = isHovered ? "show code >" : "&8show code >";
            Renderer.text(ChatLib.addColor(finalShowCode), x + width - Renderer.getStringWidth("show code >") - 2, y + height - 12)
                    .setShadow(true).draw();
        }
    }
}

File: src/main/kotlin/com/chattriggers/ctjs/engine/module/ModulesGui.kt
package com.chattriggers.ctjs.engine.module

import com.chattriggers.ctjs.minecraft.libs.ChatLib
import com.chattriggers.ctjs.minecraft.libs.renderer.Renderer
import com.chattriggers.ctjs.minecraft.libs.renderer.Text
import com.chattriggers.ctjs.minecraft.listeners.MouseListener
import com.chattriggers.ctjs.minecraft.wrappers.Client
import gg.essential.universal.UMatrixStack
import gg.essential.universal.UScreen

object ModulesGui : UScreen() {
    private val window = object {
        var title = Text("Modules").setScale(2f).setShadow(true)
        var exit = Text(ChatLib.addColor("&cx")).setScale(2f)
        var height = 0f
        var scroll = 0f
    }

    init {
        MouseListener.registerScrollListener { _, _, delta ->
            window.scroll += delta / 10
        }
    }

    //#if MC<=11202
    override fun doesGuiPauseGame() = false
    //#else
    //$$ override fun isPauseScreen() = false
    //#endif

    override fun onDrawScreen(matrixStack: UMatrixStack, mouseX: Int, mouseY: Int, partialTicks: Float) {
        super.onDrawScreen(matrixStack, mouseX, mouseY, partialTicks)
        Renderer.pushMatrix()

        val middle = Renderer.screen.getWidth() / 2f
        val width = (Renderer.screen.getWidth() - 100f).coerceAtMost(500f)

        Renderer.drawRect(
            0f,
            0f,
            Renderer.screen.getWidth().toFloat(),
            Renderer.screen.getHeight().toFloat(),
            0x50000000
        )

        if (-window.scroll > window.height - Renderer.screen.getHeight() + 20)
            window.scroll = -window.height + Renderer.screen.getHeight() - 20
        if (-window.scroll < 0) window.scroll = 0f

        if (-window.scroll > 0) {
            Renderer.drawRect(Renderer.screen.getWidth() - 20f, Renderer.screen.getHeight() - 20f, 20f, 20f, 0xaa000000)
            Renderer.drawString("^", Renderer.screen.getWidth() - 12f, Renderer.screen.getHeight() - 12f)
        }

        Renderer.drawRect(middle - width / 2f, window.scroll + 95f, width, window.height - 90, 0x50000000)

        Renderer.drawRect(middle - width / 2f, window.scroll + 95f, width, 25f, 0xaa000000)
        window.title.draw(middle - width / 2f + 5, window.scroll + 100f)
        window.exit.setString(ChatLib.addColor("&cx")).draw(middle + width / 2f - 17, window.scroll + 99f)

        window.height = 125f
        ModuleManager.cachedModules.forEach {
            window.height += it.draw(middle - width / 2f, window.scroll + window.height, width)
        }

        Renderer.popMatrix()
    }

    override fun onMouseClicked(mouseX: Double, mouseY: Double, mouseButton: Int) {
        super.onMouseClicked(mouseX, mouseY, mouseButton)

        val width = (Renderer.screen.getWidth() - 100f).coerceAtMost(500f)

        if (mouseX > Renderer.screen.getWidth() - 20 && mouseY > Renderer.screen.getHeight() - 20) {
            window.scroll = 0f
        } else if (mouseX > Renderer.screen.getWidth() / 2f + width / 2f - 25 && mouseX < Renderer.screen.getWidth() / 2f + width / 2f
            && mouseY > window.scroll + 95 && mouseY < window.scroll + 120
        ) {
            Client.currentGui.close()
        } else {
            ModuleManager.cachedModules.toList().forEach {
                it.click(mouseX, mouseY, width)
            }
        }
    }
}
